test/temporary_directory_fixture.hpp

```diff
@@ -0,0 +1,100 @@
+// Copyright 2018, Bosch Software Innovations GmbH.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef TEMPORARY_DIRECTORY_FIXTURE_HPP_
+#define TEMPORARY_DIRECTORY_FIXTURE_HPP_
+
+#include <gmock/gmock.h>
+
+#include <string>
+
+#ifdef _WIN32
+# include <direct.h>
+# include <Windows.h>
+#else
+# include <unistd.h>
+# include <sys/types.h>
+# include <dirent.h>
+#endif
+
+
+class TemporaryDirectoryFixture : public ::testing::Test
+{
+public:
+  TemporaryDirectoryFixture()
+  {
+    char template_char[] = "tmp_test_dir.XXXXXX";
+#ifdef _WIN32
+    char temp_path[255];
+    GetTempPathA(255, temp_path);
+    _mktemp_s(template_char, strnlen(template_char, 20) + 1);
+    temporary_dir_path_ = std::string(temp_path) + std::string(template_char);
+    _mkdir(temporary_dir_path_.c_str());
+#else
+    char * dir_name = mkdtemp(template_char);
+    temporary_dir_path_ = dir_name;
+#endif
+  }
+
+  ~TemporaryDirectoryFixture() override
+  {
+    remove_directory_recursively(temporary_dir_path_);
+  }
+
+  void remove_directory_recursively(const std::string & directory_path)
+  {
+#ifdef _WIN32
+    // We need a string of type PCZZTSTR, which is a double null terminated char ptr
+    size_t length = strlen(directory_path.c_str());
+    TCHAR * temp_dir = new TCHAR[length + 2];
+    memcpy(temp_dir, temporary_dir_path_.c_str(), length);
+    temp_dir[length] = 0;
+    temp_dir[length + 1] = 0;  // double null terminated
+
+    SHFILEOPSTRUCT file_options;
```

Comment:
```c++
    SHFILEOPSTRUCT file_options{
      .hwnd = nullptr,
      .wFunc =...
    };
```

would be preferable, see https://en.cppreference.com/w/c/language/struct_initialization